### PR TITLE
Improve retry logic

### PR DESCRIPTION
**Pull request by whites11**
Towards: https://github.com/ClickHouse/terraform-provider-clickhouse/issues/133

This PR:

- Replaces own code with backoff library for backoff type use cases
- Add automatic retry of 429 and 5xx errors

as a side effect this PR makes end to end tests dramatically more reliable

### CODE REVIEW

File: pkg/internal/api/client.go [hunk 1]
```diff
@@ -219,6 +227,33 @@ func (c *ClientImpl) CreateService(s Service) (*Service, string, error) {
 	return &serviceResponse.Result.Service, serviceResponse.Result.Password, nil
 }
 
+func (c *ClientImpl) WaitForServiceState(serviceId string, stateChecker func(string) bool, maxWaitSeconds int) error {
+	// Wait until service is in desired status
```
Comment (nikita-vanyasin) [inline pkg/internal/api/client.go:231]:
```suggestion
	// Wait until service is in desired state
```

File: pkg/internal/api/client.go [hunk 2]
```diff
@@ -219,6 +227,33 @@ func (c *ClientImpl) CreateService(s Service) (*Service, string, error) {
...
+	checkStatus := func() error {
```
Comment (nikita-vanyasin) [inline pkg/internal/api/client.go:232]:
```suggestion
	checkState := func() error {
```

File: pkg/internal/api/client.go [hunk 3]
```diff
@@ -219,6 +227,33 @@ func (c *ClientImpl) CreateService(s Service) (*Service, string, error) {
...
+		}
+
+		return fmt.Errorf("service %s is in state %s", serviceId, service.State)
+	}
+
+	if maxWaitSeconds < 5 {
+		maxWaitSeconds = 5
+	}
+
+	err := backoff.Retry(checkStatus, backoff.WithMaxRetries(backoff.NewConstantBackOff(5*time.Second), uint64(maxWaitSeconds/5)))
```
Comment (nikita-vanyasin) [inline pkg/internal/api/client.go:249]:
```suggestion
	err := backoff.Retry(checkState, backoff.WithMaxRetries(backoff.NewConstantBackOff(5*time.Second), uint64(maxWaitSeconds/5)))
```

File: pkg/internal/api/client.go [hunk 4]
```diff
@@ -300,27 +335,13 @@ func (c *ClientImpl) UpdateServicePassword(serviceId string, u ServicePasswordUp
...
-}
-
 func (c *ClientImpl) DeleteService(serviceId string) (*Service, error) {
 	service, err := c.GetService(serviceId)
 	if err != nil {
 		return nil, err
 	}
 
-	if service.State != "stopped" && service.State != "stopping" {
+	if service.State != StatusStopped && service.State != StatusStopping {
```
Comment (nikita-vanyasin) [inline pkg/internal/api/client.go:344]: Does it make sense to rename these constants from StatusXXX to StateXXX? To make sure naming is consistent with what is shown in CP?

**Review by nikita-vanyasin — APPROVED** [review]

File: pkg/internal/api/client.go (hunk 4 above)
Comment (whites11) [reply pkg/internal/api/client.go:344]: ok will rename everything.

**Pull request merged**